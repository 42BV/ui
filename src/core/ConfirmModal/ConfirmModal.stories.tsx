import { storiesOf } from '@storybook/react';
import ConfirmModal from './ConfirmModal';
import React, { useState } from 'react';
import { action } from '@storybook/addon-actions';
import {
  ButtonDropdown,
  DropdownItem,
  DropdownMenu,
  DropdownToggle
} from 'reactstrap';

const disclaimer = (
  <>
    <p className="mt-3">
      <strong>Note:</strong> In most cases, you&apos;ll want to use the{' '}
      <code>ConfirmButton</code> component instead. ConfirmButton offers an
      easy-to-use button that triggers this dialog.
    </p>
    <p>
      When it is not possible to trigger the action through a button, such as
      when using a ButtonDropdown, you can use this component directly instead.
    </p>
  </>
);

storiesOf('core|ConfirmModal', module)
  .addParameters({ component: ConfirmModal })
  .add('in dropdown', () => {
    const [isDropdownMenuOpen, setDropdownMenuOpen] = useState(false);
    const [isConfirmModalOpen, setConfirmModalOpen] = useState(false);

    return (
      <div className="text-center mt-3">
        <ConfirmModal
          isOpen={isConfirmModalOpen}
          onClose={() => setConfirmModalOpen(false)}
          onSave={() => {
            setConfirmModalOpen(false);
            action('confirm clicked')();
          }}
          modalText="Are you sure to delete this user? This operation cannot be undone!"
        />
        <ButtonDropdown
          isOpen={isDropdownMenuOpen}
          toggle={() => setDropdownMenuOpen(!isDropdownMenuOpen)}
        >
          <DropdownToggle caret>Actions</DropdownToggle>
          <DropdownMenu>
            <DropdownItem onClick={() => setConfirmModalOpen(true)}>
              Delete user
            </DropdownItem>
          </DropdownMenu>
        </ButtonDropdown>

        {disclaimer}
      </div>
    );
  })
  .add('with custom text', () => {
    const [isDropdownMenuOpen, setDropdownMenuOpen] = useState(false);
    const [isConfirmModalOpen, setConfirmModalOpen] = useState(false);

    return (
      <div className="text-center mt-3">
        <ConfirmModal
          isOpen={isConfirmModalOpen}
          onClose={() => setConfirmModalOpen(false)}
          onSave={() => {
            setConfirmModalOpen(false);
            action('Confirm clicked')();
          }}
          label="PLEASE SAY YES"
          modalText={
            <p>
              Are you sure you want to <strong>delete</strong> all users?
            </p>
          }
          confirmText="YES"
          cancelText="NO"
        />
        <ButtonDropdown
          isOpen={isDropdownMenuOpen}
          toggle={() => setDropdownMenuOpen(!isDropdownMenuOpen)}
        >
          <DropdownToggle caret>Actions - custom text</DropdownToggle>
          <DropdownMenu>
            <DropdownItem onClick={() => setConfirmModalOpen(true)}>
              Delete all users
            </DropdownItem>
          </DropdownMenu>
        </ButtonDropdown>

        {disclaimer}
      </div>
    );
  });
