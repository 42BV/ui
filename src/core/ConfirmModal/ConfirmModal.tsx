import React from 'react';
import { t } from '../../utilities/translation/translation';
import { Modal } from '../Modal/Modal';
import { Button } from '../Button/Button';

type Props = {
  /**
   * The text to show in the heading of the modal. Defaults to "Confirmation"
   *
   * @default Confirmation
   */
  label?: string;

  /**
   * The text to show as the cancel button's text, defaults to "Cancel"
   *
   * @default Cancel
   */
  cancelText?: string;

  /**
   * The text to show as the ok button's text, defaults to "OK"
   *
   * @default OK
   */
  confirmText?: string;

  /**
   * Function that gets called when the modal is dismissed or closed via the close button
   *
   */
  onClose: () => void;

  /**
   * Function that gets called when the user clicks the save button
   */
  onSave: () => void;

  /**
   * Text which is rendered inside the modal
   */
  modalText: React.ReactNode;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

/**
 * ConfirmModal offers the user a confirmation dialog before performing an operation.
 * The main use case is a delete action, which we want the user to confirm before proceeding.
 *
 * Note: In most cases, you'll want to use {@link ConfirmButton} instead.
 * That component offers an easy-to-use button that triggers this dialog.
 * However, if it is not possible to trigger the action through a button,
 * such as when using a ButtonDropdown, you can use this component directly instead.
 */
export function ConfirmModal({
  onClose,
  onSave,
  label,
  cancelText,
  confirmText,
  modalText,
  className
}: Props) {
  return (
    <Modal
      onClose={onClose}
      className={className}
      header={t({
        overrideText: label,
        key: 'ConfirmButton.MODAL_HEADER',
        fallback: 'Confirmation'
      })}
      footer={
        <>
          <Button
            className="ms-1"
            color="secondary"
            icon="close"
            onClick={onClose}
          >
            {t({
              overrideText: cancelText,
              key: 'ConfirmButton.CANCEL',
              fallback: 'Cancel'
            })}
          </Button>
          <Button
            className="ms-1"
            color="primary"
            icon="check"
            onClick={onSave}
          >
            {t({
              overrideText: confirmText,
              key: 'ConfirmButton.CONFIRM',
              fallback: 'Confirm'
            })}
          </Button>
        </>
      }
    >
      {modalText}
    </Modal>
  );
}
