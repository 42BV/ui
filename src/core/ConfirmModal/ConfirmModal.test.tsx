import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { ConfirmModal } from './ConfirmModal';

describe('Component: ConfirmModal', () => {
  describe('ui', () => {
    test('default texts', () => {
      render(
        <ConfirmModal onClose={() => undefined} onSave={() => undefined}>
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        </ConfirmModal>
      );

      expect(document.body.lastChild).toMatchSnapshot();
    });

    test('custom texts', () => {
      render(
        <ConfirmModal
          onClose={() => undefined}
          onSave={() => undefined}
          cancelText="NO"
          confirmText="YES"
          label="Perform dangerous action"
        >
          <p>
            Are you sure you want to <strong>delete</strong> all users in the
            database?
          </p>
        </ConfirmModal>
      );

      expect(screen.queryByText('Confirmation')).not.toBeInTheDocument();
      expect(
        screen.queryByText('Perform dangerous action')
      ).toBeInTheDocument();
      expect(screen.getAllByRole('button')[1]).toHaveTextContent('cancelNO');
      expect(screen.getAllByRole('button')[2]).toHaveTextContent('saveYES');
      expect(document.body.lastChild).toMatchSnapshot();
    });
  });

  describe('events', () => {
    type Props = { text: string; isOpen?: boolean };

    function setup({ text, isOpen }: Props) {
      const onCloseSpy = jest.fn();
      const onSaveSpy = jest.fn();

      const props = {
        isOpen: isOpen ?? false,
        onClose: onCloseSpy,
        onSave: onSaveSpy,
        label: 'Are you sure you want a ConfirmAction?',
        children: text
      };

      const { container } = render(<ConfirmModal {...props} />);

      return { container, onCloseSpy, onSaveSpy };
    }

    it('should open the OpenCloseModal when the ConfirmModal is opened', () => {
      setup({
        text: 'Delete all data in the database?',
        isOpen: true
      });

      expect(
        screen.queryByText('Delete all data in the database?')
      ).toBeInTheDocument();
    });

    it('should call onClose when the Modal is closed', () => {
      const { onCloseSpy, onSaveSpy } = setup({
        text: 'Delete this user and all his data?',
        isOpen: true
      });

      fireEvent.click(screen.getByLabelText('Close'));

      expect(onCloseSpy).toHaveBeenCalledTimes(1);
      expect(onSaveSpy).toHaveBeenCalledTimes(0);
    });

    it('should call onClose when the cancel button is clicked', () => {
      const { onCloseSpy, onSaveSpy } = setup({
        text: 'Delete this user and all his data?',
        isOpen: true
      });

      fireEvent.click(screen.getByText('cancel'));

      expect(onCloseSpy).toHaveBeenCalledTimes(1);
      expect(onSaveSpy).toHaveBeenCalledTimes(0);
    });

    it('should call onSave when the Modal is confirmed', () => {
      const { onCloseSpy, onSaveSpy } = setup({
        text: 'Delete this user and all his data?',
        isOpen: true
      });

      fireEvent.click(screen.getByText('save'));

      expect(onCloseSpy).toHaveBeenCalledTimes(0);
      expect(onSaveSpy).toHaveBeenCalledTimes(1);
    });
  });
});
