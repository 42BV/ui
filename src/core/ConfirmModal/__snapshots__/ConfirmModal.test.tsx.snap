// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ConfirmModal ui custom texts 1`] = `
<div
  style="position: relative; z-index: 1050; display: block;"
  tabindex="-1"
>
  <div
    class=""
  >
    <div
      aria-modal="true"
      class="modal fade"
      role="dialog"
      style="display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Perform dangerous action
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <p>
              Are you sure you want to 
              <strong>
                delete
              </strong>
               all users in the database?
            </p>
          </div>
          <div
            class="modal-footer"
          >
            <span
              class="button d-inline-block ms-1 secondary"
            >
              <button
                class="btn btn-secondary btn-md"
                type="button"
              >
                <div
                  class="d-flex justify-content-center align-items-center"
                >
                  <i
                    class="icon material-icons material-icons__filled button-icon me-2"
                  >
                    close
                  </i>
                  NO
                </div>
              </button>
            </span>
            <span
              class="button d-inline-block ms-1 primary"
            >
              <button
                class="btn btn-primary btn-md"
                type="button"
              >
                <div
                  class="d-flex justify-content-center align-items-center"
                >
                  <i
                    class="icon material-icons material-icons__filled button-icon me-2"
                  >
                    check
                  </i>
                  YES
                </div>
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="modal-backdrop fade"
    />
  </div>
</div>
`;

exports[`Component: ConfirmModal ui default texts 1`] = `
<div
  style="position: relative; z-index: 1050; display: block;"
  tabindex="-1"
>
  <div
    class=""
  >
    <div
      aria-modal="true"
      class="modal fade"
      role="dialog"
      style="display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Confirmation
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <p>
              Are you sure you want to 
              <strong>
                delete
              </strong>
               the user?
            </p>
          </div>
          <div
            class="modal-footer"
          >
            <span
              class="button d-inline-block ms-1 secondary"
            >
              <button
                class="btn btn-secondary btn-md"
                type="button"
              >
                <div
                  class="d-flex justify-content-center align-items-center"
                >
                  <i
                    class="icon material-icons material-icons__filled button-icon me-2"
                  >
                    close
                  </i>
                  Cancel
                </div>
              </button>
            </span>
            <span
              class="button d-inline-block ms-1 primary"
            >
              <button
                class="btn btn-primary btn-md"
                type="button"
              >
                <div
                  class="d-flex justify-content-center align-items-center"
                >
                  <i
                    class="icon material-icons material-icons__filled button-icon me-2"
                  >
                    check
                  </i>
                  Confirm
                </div>
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="modal-backdrop fade"
    />
  </div>
</div>
`;
